src/tree/updater_quantile_hist.cc

```diff
@@ -740,33 +761,19 @@ void QuantileHistMaker::Builder<GradientSumT>::InitSampling(const std::vector<Gr
       const size_t ibegin = tid * discard_size;
       const size_t iend = (tid == (nthread - 1)) ?
                           info.num_row_ : ibegin + discard_size;
+      constexpr uint64_t base = 16807;
+      uint64_t initial_seed_th = simple_skip(discard_size * tid, initial_seed, base);
+      std::linear_congruential_engine<std::uint_fast64_t, base, 0, (1ULL << 63ULL)> eng(initial_seed_th);
 
-      rnds[tid].discard(discard_size * tid);
       for (size_t i = ibegin; i < iend; ++i) {
-        if (gpair[i].GetHess() >= 0.0f && rnds[tid]() < coin_flip_border) {
-          p_row_indices_used[ibegin + row_offsets_used[tid]++] = i;
-        } else {
+        if (!(gpair[i].GetHess() >= 0.0f && coin_flip(eng))) {
           p_row_indices_unused[ibegin + row_offsets_unused[tid]++] = i;
```

Comment:
Since you have 0 gradient for ignored samples, is it still necessary for the partitioner to be aware of sampling?

Reply:
@trivialfis sorry for long response (was trying to fix cpp/python tests).
Practitioner still has to be aware about 'unused' rows to do `UpdatePredictionCache` correctly.